File: sds/category/그래프1/집합의표현1717/P1717.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class P1717 {
    static int N,M;
    static int[] nums;
    public static void main(String[] args) throws Exception{
        System.setIn(new FileInputStream("./input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        nums = new int[N+1];
        for (int i = 0; i < N+1; i++) {
            nums[i] = i;
        }
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int type = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            if(type==0){
                union(a, b);
            }else{
                int parA = find(a);
                int parB = find(b);
                if(parA==parB) sb.append("YES"+"\n");
                else sb.append("NO"+"\n");
            }
        }
        System.out.println(sb);
    }
    static int find(int target){
        if(target==nums[target]) return target;
        return nums[target] = find(nums[target]);
    }
    static void union(int tA, int tB){
        int parA = find(tA);
        int parB = find(tB);
        if( parA ==parB) return;
        nums[parB] = parA;
    }
}

File: baekjun/basic/P1717/P1717.py
import sys
sys.setrecursionlimit(10**6)
sys.stdin = open("E:/algorithm/baekjun/basic/P1717/input.txt")
N,M = map(int,sys.stdin.readline().split())
par = [i for i in range(N+1)]

def find(a):
    par_a = par[a]
    if par_a!=a:
        par[a] = find(par_a)
        return par[a]
    else:
        return a

def union(a,b):
    par_a = find(a)
    par_b = find(b)
    if par_a!=par_b:
        par[par_a] = par_b

for _ in range(M):
    t,i,j = map(int,sys.stdin.readline().split())
    if t==0:
        union(i,j)
    else:
        par_a = find(i)
        par_b = find(j)
        if par_a == par_b:
            sys.stdout.write("YES"+"\n")
        else:
            sys.stdout.write("NO"+"\n")
